homeassistant/components/tahoma/tahoma_device.py

```diff
@@ -0,0 +1,169 @@
+        attr = {
+            "ui_class": self.device.ui_class,
+            "widget": self.device.widget,
+            "controllable_name": self.device.controllable_name,
+        }
+
+        if self.has_state(CORE_RSSI_LEVEL_STATE):
+            attr[ATTR_RSSI_LEVEL] = self.select_state(CORE_RSSI_LEVEL_STATE)
+
+        if self.has_state(CORE_BATTERY_STATE):
+            battery_state = self.select_state(CORE_BATTERY_STATE)
+
+            if battery_state == STATE_BATTERY_FULL:
+                battery_state = 100
+            elif battery_state == STATE_BATTERY_NORMAL:
+                battery_state = 75
+            elif battery_state == STATE_BATTERY_LOW:
+                battery_state = 25
+            elif battery_state == STATE_BATTERY_VERY_LOW:
+                battery_state = 10
+
+            attr[ATTR_BATTERY_LEVEL] = battery_state
+
+        if self.select_state(CORE_SENSOR_DEFECT_STATE) == STATE_DEAD:
+            attr[ATTR_BATTERY_LEVEL] = 0
+
+        if self.device.attributes:
+            for attribute in self.device.attributes:
+                attr[attribute.name] = attribute.value
+
+        if self.device.states:
+            for state in self.device.states:
+                if "State" in state.name:
+                    attr[state.name] = state.value
+
+        return attr
+
+    @property
+    def device_info(self) -> Dict[str, Any]:
+        """Return device registry information for this entity."""
+        manufacturer = self.select_state(CORE_MANUFACTURER_NAME_STATE) or "Somfy"
+        model = self.select_state(CORE_MODEL_STATE) or self.device.widget
+
+        return {
+            "identifiers": {(DOMAIN, self.unique_id)},
+            "manufacturer": manufacturer,
+            "name": self.name,
+            "model": model,
+            "sw_version": self.device.controllable_name,
+        }
+
+    def select_command(self, *commands: str) -> Optional[str]:
+        """Select first existing command in a list of commands."""
+        existing_commands = self.device.definition.commands
+        return next((c for c in commands if c in existing_commands), None)
+
+    def has_command(self, *commands: str) -> bool:
+        """Return True if a command exists in a list of commands."""
+        return self.select_command(*commands) is not None
+
+    def select_state(self, *states) -> Optional[str]:
+        """Select first existing active state in a list of states."""
+        if self.device.states:
+            return next(
+                (
+                    state.value
+                    for state in self.device.states
+                    if state.name in list(states)
+                ),
+                None,
+            )
+        return None
+
+    def has_state(self, *states: str) -> bool:
+        """Return True if a state exists in self."""
+        return self.select_state(*states) is not None
+
+    async def async_execute_command(self, command_name: str, *args: Any):
```

Comment:
I recommend moving all device api helpers to a separate class that represents the device and give the entity access to an instance of that class. Mixing in the device api in the entity api often leads to bad patterns.

Reply:
@tetienne, @vlebourl would you know how to do this? I thought about creating a `OverkizDevice` class, however in almost all device specific features we need the coordinator data / client. Not sure how this would work with inheritance in Python.

Reply:
Done!